### PR TITLE
[AST] Strip InOutType when cloning a parameter.

### PR DESCRIPTION
**Pull request by DougGregor**
Fixes rdar://problem/34789779, where initializer inheritance was breaking
in the presence of 'inout' parameters.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -4197,12 +4197,20 @@ ParamDecl::ParamDecl(Specifier specifier,
            "'var' cannot appear on parameters; you meant 'inout'");
 }
 
+// Remove the top-level inout type, if there is one.
+static Type removeInOutType(Type type) {
```
Comment (CodaFi) [inline lib/AST/Decl.cpp]: `getInOutObjectType()` already handles this and I would rather keep usages of `getAs<InOutType>()` centralized.

Comment (DougGregor) [reply lib/AST/Decl.cpp]: Ugh, how did I miss `getInOutObjectType()`? Search fail, thank you!

**Comment by DougGregor** [comment]
In my haste to not get this auto-merged, I busted it. Opened a new PR as https://github.com/apple/swift/pull/12272

**Pull request closed without merging**